accepted/2024/package-update-command.md

```diff
@@ -0,0 +1,238 @@
+
+#### dotnet package update
+
+The `dotnet package update` command will update all the NuGet dependencies to the latest versions.
+
+```CLI
+C:\> dotnet package update --help
+dotnet package update [<PROJECT>|<SOLUTION>] [--vulnerable] [--mode <MODE>] [-v|--verbosity <LEVEL>] [--json] 
+```
+
+#### --vulnerable
+
+This flag will try to update only packages that have direct or transitive vulnerabilities. The remediation is calculated with an implicit dotnet audit to then apply directly to a resulting package graph. It can add packages, remove packages, and update packages depending on the problem it's attempting to resolve. It does not take into consideration downgrading to a compatible version if a higher one has already been specified.
+The algorithm of the fixing process will have the following steps:
+
+1. Identify  vulnerabilities in the graph
+1. Distinguish which are top level, transitive level, etc.,
+1. Resolve any compatibility conflicts
+1. Apply the patch/update/fix
+1. Restore dependencies
+1. Verify that issues were resolved  . Provide the report to the user.
+1. If the issues were not resolved or were partially resolved, we will still commit the fixes and notify users about what was and was not fixed and what are the recommended manual steps with links to documentation.
+
+Example:
+
+```CLI
+C:\ContosoApp\> dotnet package update --vulnerable
+
+Fixing vulnernable packages in ContosoUniversity.sln
+ ContosoLibrary:
+  Upgrading UmbracoForms 8.4.1 to 8.7.1
+
+ ContosoApp:
+  Updated UmbracoForms 8.4.0 to 8.7.1
+    
+Fixed 2 packages in 36 scanned packages.
+```
+
+Another option for the output with more information (will review with CLI team to decide which one to pick):
+
+```CLI
+C:\> dotnet package update --vulnerable [<SOLUTION_PATH>|<PROJECT_PATH_>]
+
+Fetching package metadata from: 'https://api.nuget.org/v3/index.json'
+Loaded 23 security advisories from 'https://api.nuget.org/v3/index.json'
+Scanning ContosoUniversity.sln (36 NuGet dependencies)
+
+error: Vulnernable packages found!
+[net5.0]:
+Top-level Package     Requested Resolved Severity Advisory URL
+> UmbracoForms      8.4.1  8.4.1  Moderate https://github.com/advisories/GHSA-8m73-w2r2-6xxj
+
+Transitive Package  Resolved Severity Advisory URL
+> Microsoft.Data.OData 5.2.0  Moderate https://github.com/advisories/GHSA-mv2r-q4g5-j8q5
+
+Found 1 top-level Moderate severity vulnerability & 1 transtive Moderate severity vulnerability package(s) in 36 scanned packages.
+
+Fixing vulnernable packages in ContosoUniversity.sln
+ Upgrading UmbracoForms 8.4.1 to 8.7.1
+
+Fixed 2 packages in 36 scanned packages.
+```
+
+#### --mode
+
+Priority: P1 for `promote`; P2 for `closest` and switching the default to `closest`.
+
+| Mode | Explanation |
+|------|-------------|
+| `promote` | the algorithm will always promote transitive packages to top level. This is an easy to implement feature that we will release in MVP. |
```

Comment:
I'm not sure where in the document it should be written, but while talking to an internal partner, they suggested that packages that are promoted to top level in order to remove a vulnerable package should have some additional metadata to signal this. For example `<PackageReference Include="SomePackage" Version="1.2.3" PromotedBecauseOfVulnerability="true" />`.  A much, much better name is needed though, this is just an example.

If in the future the top level package that originally caused SomePackage to be included in the package graph has a new version that removes SomePackage from the package graph, this extra metadata will allow a future improvement to `dotnet update package` to detect that the PackageReference can now be removed.

I think this is a great idea for projects that aren't using CPM with transitive pinning. Although it's also a signal that transitive pinning without CPM would be a useful feature, but very much out of scope for this feature 😁